File: proposed-projects/round-trip-experiment/reverse/20260909T150124Z_a7cd2ee9/20260910T084451Z_b3565ede/ported_implementation/gbnf/utils/errors/build_error_position.py
MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> list[str]:
    if src == "":
        return ["No input provided"]
    lines = src.split("\n")

    line_idx = 0
    while (
        line_idx < len(lines)
        and lines[line_idx]
        and pos > len(lines[line_idx]) - 1
        and pos < len(src)
    ):
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show = lines[
        max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)) : line_idx + 1
    ]

    return [*lines_to_show, f"{' ' * max(0, pos)}^"]

File: proposed-projects/round-trip-experiment/reverse/20260909T013603Z_054a84ca/20260910T052226Z_c9b9c06d/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  while (lines[lineIdx] && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  for (let i = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)); i <= lineIdx; i++) {
    // Out-of-range lines are `undefined`; joining renders them as an empty line.
    linesToShow.push(lines[i] ?? '');
  }

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};
